base/docs/Docs.jl

```diff
@@ -285,41 +279,59 @@ function objdoc(meta, def)
     end
 end
 
-fexpr(ex) = isexpr(ex, :function, :(=)) && isexpr(ex.args[1], :call)
+function vardoc(meta, def, name)
+    quote
+        @init
+        f = $(esc(def))
+        # @var isn't found – bug?
+        doc!(Meta.@var($(esc(name))), $(mdify(meta)))
+        f
+    end
+end
 
 function docm(meta, def)
-    # Quote, Unblock and Macroexpand
-    # * Always do macro expansion unless it's a quote (for consistency)
-    # * Unblock before checking for Expr(:quote) to support `->` syntax
-    # * Unblock after macro expansion to recognize structures of
-    #   the generated AST
-    def′ = unblock(def)
-    if !isexpr(def′, :quote)
-        def = macroexpand(def)
-        def′ = unblock(def)
-    elseif length(def′.args) == 1 && isexpr(def′.args[1], :macrocall)
-        # Special case for documenting macros after definition with
-        # `@doc "<doc string>" :@macro` or
-        # `@doc "<doc string>" :(str_macro"")` syntax.
-        #
-        # Allow more general macrocall for now unless it causes confusion.
-        return objdoc(meta, namify(def′.args[1]))
-    end
-    isexpr(def′, :macro) && return namedoc(meta, def, symbol("@", namify(def′)))
-    isexpr(def′, :type) && return typedoc(meta, def, namify(def′.args[2]))
-    isexpr(def′, :bitstype) && return namedoc(meta, def, def′.args[2])
-    isexpr(def′, :abstract) && return namedoc(meta, def, namify(def′))
-    isexpr(def′, :module) && return namedoc(meta, def, def′.args[2])
-    fexpr(def′) && return funcdoc(meta, def)
-    return objdoc(meta, def)
+    @match def begin
+        :(@m_)      -> return objdoc(meta, m)
+          m_""      -> return objdoc(meta, m)
+        (@var x_)   -> vardoc(meta, def, x)
+    end
+    def = macroexpand(def)
+    @match def begin
+        (f_(__) = _)          -> funcdoc(meta, def)
+        function f_(__) _ end -> funcdoc(meta, def)
+        function f_ end       ->  objdoc(meta, def)
+        macro m_(__) _ end    -> namedoc(meta, def, symbol("@", m))
+
+        type T_ _ end         -> typedoc(meta, def, namify(T))
+        immutable T_ _ end    -> typedoc(meta, def, namify(T))
+        (abstract T_)         -> namedoc(meta, def, namify(T))
+        (bitstype _ T_)       -> namedoc(meta, def, namify(T))
+        (typealias T_ _)      ->  objdoc(meta, def)
```

Comment:
@one-more-minute, from [this comment](https://github.com/JuliaLang/julia/issues/12102#issuecomment-120534943), can ``(typealias T_ _)`` be rewritten as ``(typealias _ _)``?